warehouse/manage/forms.py

```diff
@@ -238,12 +251,59 @@ def validate_token_scope(self, field):
 
         if scope_kind != "project":
             raise wtforms.ValidationError(f"Unknown token scope: {scope}")
-        if scope_value not in self.project_names:
-            raise wtforms.ValidationError(
-                f"Unknown or invalid project name: {scope_value}"
+        for project in self.all_projects:
+            if scope_value == project.normalized_name:
+                return
+        raise wtforms.ValidationError(f"Unknown or invalid project name: {scope_value}")
+
+    def validate_releases(self, field):
+        release = field.data
+        try:
+            releases = release.split(".")
+            for val in releases:
+                int(val)
+        except ValueError:
+            raise wtforms.validators.ValidationError("Invalid release")
+
+        for project in self.all_projects:
+            for version in project.all_versions:
+                if version[0] == release:
+                    raise wtforms.validators.ValidationError("Invalid release")
+
+    def validate_expiration(self, field):
+        expiration = field.data
+        expiration = datetime.strptime(expiration, "%Y-%m-%dT%H:%M")
+        d = datetime.now()
+        tz = pytz.timezone("GMT")  # GMT for POC, ideally would be user's local timezone
+        tz_aware = tz.localize(d)
+        expiration_aware = tz.localize(expiration)
+
+        if expiration_aware > tz_aware + timedelta(days=365):
+            raise wtforms.validators.ValidationError(
+                "Expiration cannot be greater than one year"
+            )
+        if expiration_aware < tz_aware:
+            raise wtforms.validators.ValidationError(
+                "Expiration must be after the current time"
             )
 
-        self.validated_scope = {"projects": [scope_value]}
+        expiration = datetime.strftime(expiration, "%Y-%m-%dT%H:%M")
+
+    def validate(self):
+        res = super().validate()
+        if not isinstance(self.validated_scope, str):
+            self.validated_scope = {
+                "expiration": self.expiration.data,
+                "projects": [
+                    {
+                        "project-name": self.token_scope.data,
```

Comment:
Could be misunderstanding here, but won't this embed the full `scope:` prefix into the `project-name`? I think we need to parse it out first.